File: greedy/python/police_and_thieves.py
class Solution:
    def catchThieves(self, arr, n, k): 
        thieves_index = []
        police_index = []
        
        # Obtém os índices para ladrões e policiais
        for i in range(n):
            if arr[i] == 'T':
                thieves_index.append(i)
            elif arr[i] == 'P':
                police_index.append(i)
        
        t = 0
        p = 0
        ans = 0
        # Enquanto houver ladrões ou policiais restantes
        while t < len(thieves_index) and p < len(police_index):
            
            dist = abs(thieves_index[t] - police_index[p])
            if dist <= k:
                ans += 1
                t += 1
                p += 1
            elif thieves_index[t] < police_index[p]:
                t += 1
            else:
                p += 1
        
        return ans
    
s = Solution()
print(s.catchThieves(['P', 'T', 'T', 'P', 'T'], 5, 1))
print(s.catchThieves(['T', 'T', 'P', 'P', 'T', 'P'], 6, 2))

File: greedy/java/src/PoliceAndThieves.java
import java.util.ArrayList;
import java.util.List;

public class PoliceAndThieves {
	public static void main(String[] args) {
		System.out.println(catchThieves(new char[] {'P', 'T', 'T', 'P', 'T'}, 5, 1));
		System.out.println(catchThieves(new char[] {'T', 'T', 'P', 'P', 'T', 'P'}, 6, 2));
	}
	
    public static int catchThieves(char[] arr, int n, int k) {
        List<Integer> thievesIndex = new ArrayList<>();
        List<Integer> policeIndex = new ArrayList<>();
        
        // Obtém os índices para ladrões e policiais
        for (int i = 0; i < n; i++) {
            if (arr[i] == 'T')
                thievesIndex.add(i);
            else if (arr[i] == 'P')
                policeIndex.add(i);
        }
        
        int t = 0, p = 0, ans = 0;
        // Enquanto houver ladrões ou policiais restantes
        while (t < thievesIndex.size() && p < policeIndex.size()) {
            
            int dist = Math.abs(thievesIndex.get(t) - policeIndex.get(p));
            if (dist <= k) {
                ans++;
                t++;
                p++;
            }
            else if (thievesIndex.get(t) < policeIndex.get(p)) {
                t++;
            }
            else {
                p++;
            }
        }
        
        return ans;
    }
}
